Sources/Basics/Cancellator.swift

```diff
@@ -24,10 +28,77 @@ public class Cancellator: Cancellable {
     private let cancelationQueue = DispatchQueue(label: "org.swift.swiftpm.cancellator", qos: .userInteractive, attributes: .concurrent)
     private let cancelling = ThreadSafeBox<Bool>(false)
 
+    private static var isSignalHandlerInstalled = false
+
     public init(observabilityScope: ObservabilityScope?) {
         self.observabilityScope = observabilityScope
     }
 
+    #if os(Windows)
+    // unfortunately this is needed for C callback handlers used by Windows shutdown handler
+    static var shared: Cancellator?
+    #endif
+
+    /// Installs signal handlers to terminate sub-processes on cancellation.
+    public func installSignalHandlers() {
+        precondition(!Self.isSignalHandlerInstalled)
+
+        #if os(Windows)
+        // Closures passed to `SetConsoleCtrlHandler` can't capture context, working around that with a global.
+        Self.shared = self
+
+        // set shutdown handler to terminate sub-processes, etc
+        _ = SetConsoleCtrlHandler({ _ in
+            // Terminate all processes on receiving an interrupt signal.
+            try? Self.shared?.cancel(deadline: .now() + .seconds(30))
+
+            // Reset the handler.
+            _ = SetConsoleCtrlHandler(nil, false)
+
+            // Exit as if by signal()
+            TerminateProcess(GetCurrentProcess(), 3)
+
+            return true
+        }, true)
+        #else
+        // trap SIGINT to terminate sub-processes, etc
+        signal(SIGINT, SIG_IGN)
+        let interruptSignalSource = DispatchSource.makeSignalSource(signal: SIGINT)
+        interruptSignalSource.setEventHandler { [weak self] in
+            // cancel the trap?
+            interruptSignalSource.cancel()
+
+            // Terminate all processes on receiving an interrupt signal.
+            try? self?.cancel(deadline: .now() + .seconds(30))
+
+            #if os(macOS) || os(OpenBSD)
+            // Install the default signal handler.
+            var action = sigaction()
+            action.__sigaction_u.__sa_handler = SIG_DFL
+            sigaction(SIGINT, &action, nil)
+            kill(getpid(), SIGINT)
+            #elseif os(Android)
+            // Install the default signal handler.
+            var action = sigaction()
+            action.sa_handler = SIG_DFL
+            sigaction(SIGINT, &action, nil)
+            kill(getpid(), SIGINT)
+            #else
+            var action = sigaction()
+            action.__sigaction_handler = unsafeBitCast(
+                SIG_DFL,
+                to: sigaction.__Unnamed_union___sigaction_handler.self
+            )
+            sigaction(SIGINT, &action, nil)
+            kill(getpid(), SIGINT)
+            #endif
+        }
+        interruptSignalSource.resume()
+        #endif
+
+        Self.isSignalHandlerInstalled = true
```

Comment:
non blocking but this is not thread safe

Reply:
I've added a lock for the whole function body to make sure that there's no race condition when installing these handlers.